app/src/main/java/net/justdave/mcstatus/MinecraftServer.java

```diff
@@ -288,33 +283,38 @@ public void query() {
 			out.write(0x00); // status ping (2nd byte)
 
 			int packetLength = readVarInt(in); // size of entire packet
+			String serverData;
 			if (packetLength < 11) {
-				Log.i(TAG, "packet length too small");
+				Log.i(TAG, String.format("%s, %s: packet length too small: %d", serverName, serverAddress, packetLength));
 				serverData = null;
-				setDescription("Invalid response from server (packet too small - server may be in the process of restarting, try again in a few seconds)");
+				setDescription("Invalid response from server (server may be in the process of restarting, try again in a few seconds)");
+				in.close();
+				out.close();
+				socket.close();
 				return;
 			}
-			in.read(); // packet type - going to ignore it because we should
-						// only ever get the one type back anyway
+			final int packetType = in.read(); // packet type - going to ignore it because we should
+											  // only ever get the one type back anyway
+			Log.d(TAG, String.format("%s, %s: packet type: %d", serverName, serverAddress, packetType));
 			int jsonLength = readVarInt(in); // size of JSON blob
+			if (jsonLength < 0){
+				in.close();
+				out.close();
+				socket.close();
+				return;
+			}
 			byte[] buffer = new byte[jsonLength + 10]; // a little more than
 														// we're expecting just
 														// to be safe
 			int bytesRead = 0;
 			do {
 				bytesRead += in.read(buffer, bytesRead, jsonLength - bytesRead);
 			} while (bytesRead < jsonLength);
-			if (bytesRead < jsonLength) {
```

Comment:
Because of the `do, while` loop condition above this statement can never be true (android studio suggested this)